### PR TITLE
Add support for different result formats to MSQ SqlStatementResource

### PR DESCRIPTION
**Pull request by adarshsanjeev**
Currently SqlStatementResource only supports results in "object" format. This adds support for these results to be written in any format specified in ResultFormat.

This can be done either by providing a format in the SQLQuery in the POST request,

```
{
    "query":"SELECT * FROM kttm_simple",
    "resultFormat": "object",
    ...
}
```

 or by providing "resultFormat" as a query parameter in the GET request.

```
http://localhost:8888/druid/v2/sql/statements/:id/results/?page=0&resultsFormat=objectLines
```

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/destination/MSQDestination.java [hunk 1]
```diff
@@ -30,5 +31,5 @@
 })
 public interface MSQDestination
 {
-  // No methods. Just a marker interface for deserialization.
+  ResultFormat getResultFormat();
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/destination/MSQDestination.java:34]: Lets document how this interface is used here.

Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/destination/MSQDestination.java:34]: hmm this seems like the destination is marked with "x" result format which is not the case in taskReports as well as DurableStorageDestination. 

Since we have the context of the query passed to the task, is this change required. 
We can just look as the taskPayload to figure out the "resultFormat" passed on the post statement always . 

WDYT ?

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/destination/MSQDestination.java]: Changed

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 2]
```diff
@@ -164,6 +165,7 @@ public Response isEnabled(@Context final HttpServletRequest request)
   @Consumes(MediaType.APPLICATION_JSON)
   public Response doPost(final SqlQuery sqlQuery, @Context final HttpServletRequest req)
   {
+    sqlQuery.getContext().put(MSQTaskQueryMaker.RESULT_FORMAT, sqlQuery.getResultFormat().toString());
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:168]: Can we add a line here as to why we are doing this.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/SqlMSQStatementResourcePostTest.java [hunk 3]
```diff
@@ -139,30 +140,32 @@ public void testMSQSelectQueryTest() throws IOException
...
   {
-
     SqlStatementResourceTest.assertExceptionMessage(
         resource.doPost(new SqlQuery(
             "select * from foo",
             null,
             false,
             false,
             false,
-            ImmutableMap.of(),
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/SqlMSQStatementResourcePostTest.java:150]: I think we should create a new sqlQuery object in the post request when you are mutating the map. WDYT?

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/SqlMSQStatementResourcePostTest.java [hunk 4]
```diff
@@ -327,12 +330,100 @@ public void testWithDurableStorage() throws IOException
...
+  @Test
+  public void testResultFormat() throws IOException
+  {
+    Map<String, Object> context = defaultAsyncContext();
+    context.put(MultiStageQueryContext.CTX_SELECT_DESTINATION, MSQSelectDestination.DURABLESTORAGE.name());
+
+    SqlStatementResult sqlStatementResult = (SqlStatementResult) resource.doPost(
+        new SqlQuery(
+            "select cnt,dim1 from foo",
+            ResultFormat.ARRAY,
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/SqlMSQStatementResourcePostTest.java:354]: Lets add ut's with all result formats.

File: sql/src/main/java/org/apache/druid/sql/http/SqlQuery.java [hunk 5]
```diff
@@ -66,7 +66,7 @@ public SqlQuery(
   )
   {
     this.query = Preconditions.checkNotNull(query, "query");
-    this.resultFormat = resultFormat == null ? ResultFormat.OBJECT : resultFormat;
+    this.resultFormat = resultFormat == null ? ResultFormat.DEFAULT_RESULT_FORMAT : resultFormat;
```
Comment (cryptoe) [inline sql/src/main/java/org/apache/druid/sql/http/SqlQuery.java:69]: Why is this change needed ?

Comment (adarshsanjeev) [reply sql/src/main/java/org/apache/druid/sql/http/SqlQuery.java:69]: This is the default format when nothing is specified at all, but it is only specified in this class. It would be better if this was clearly specified so that it could be reused (or changed in the future), so I added `DEFAULT_RESULT_FORMAT`

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 6]
```diff
@@ -115,6 +117,7 @@
 public class SqlStatementResource
 {
 
+  public static final String RESULT_FORMAT = "resultFormat";
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:120]: I think we should set this as __resultFormat ie this is set by the broker and not by the user directly. Sort of like an internal parameter. 
WDYT ?

**Review by cryptoe — APPROVED** [review]
Thanks for adding the test cases. 
Left some comments. 
Otherwise LGTM.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 7]
```diff
@@ -663,12 +666,36 @@ private MSQControllerTask getMSQControllerTaskOrThrow(String queryId, String cur
...
   }
 
+  /**
+   * Creates a new sqlQuery from the user submitted sqlQuery after performing required modifications.
+   */
+  private SqlQuery createModifiedSqlQuery(SqlQuery sqlQuery)
+  {
+    Map<String, Object> context = sqlQuery.getContext();
+    if (context.containsKey(RESULT_FORMAT)) {
+      throw InvalidInput.exception("Query context parameter [%s] is not allowed", RESULT_FORMAT);
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:676]: if we do this: https://github.com/apache/druid/pull/14571#discussion_r1285222902 then I think the error message need not change.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java (hunk 6 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:120]: Since this is a restricted parameter for the user, this should also be added here https://github.com/apache/druid/blob/2d8e0f28f36f83e091911cc7b36a57b10fd065cc/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java#L58

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java]: I think SYSTEM_CONTEXT_PARAMETERS are for those params set within MSQ itself. This is being set at the resource, so this should would always fail and the check wouldn't matter anyway since it is not set later in MSQ either.
